src/Psl/Str/replace_ci.php

```diff
@@ -12,12 +14,21 @@
  * `$replacement` (case-insensitive).
  *
  * @pure
+ *
+ * @throws Psl\Exception\InvariantViolationException if $needle is not a valid UTF-8 string.
  */
 function replace_ci(string $haystack, string $needle, string $replacement, Encoding $encoding = Encoding::UTF_8): string
 {
     if ('' === $needle || null === search_ci($haystack, $needle, 0, $encoding)) {
         return $haystack;
     }
 
-    return join(preg_split('{' . preg_quote($needle, '/') . '}iu', $haystack), $replacement);
+    /** @psalm-suppress ImpureFunctionCall */
+    $pieces = Psl\Internal\suppress(
```

Comment:
I can't seem to figure out why you need to suppress:
https://3v4l.org/toK8g

I'm not getting any warnings from PHP (even with error reporting on `E_ALL`)

Reply:
I replicated it on this example https://3v4l.org/citBH

Reply:
Ah yes I see, it's when the expression contains issues.
This also aplies to preg_replace though:
https://3v4l.org/bXP6u

So both the needle and/or haystack can fail in this case. 
The same goes for the trim functions : both the string and char_mask might fail.

I suggest adding suppress for every function and generalizing the error message so that either the needle and/or haystack may contain invalid chars?

Reply:
> So both the needle and/or haystack can fail in this case.

Indeed, I had serious doubts about the invariant message. Even about the valid UTF-8 string. What if the function failed because of another reason? I could play with [`preg_last_error`](https://www.php.net/manual/es/function.preg-last-error.php) here

> I suggest adding suppress for every function and generalizing the error message so that either the needle and/or haystack may contain invalid chars?

As I tested it, no warnings appeared when throwing an invalid UTF-8. Anyway, it may be wise to silence all kind of error and use the same `preg_last_error` function. I'll work on it!

Reply:
Well, after thinking about it, wouldn't be the same as using `Psl\Regex`, probably with the same overhead? 😕

Reply:
Here's a funny thing I'm not sure how to solve, because I don't know if the problem is on the test's data set.

If I call `Regex\split`, the following assertion fails:
```php
function replace_ci(string $haystack, string $needle, string $replacement, Encoding $encoding = Encoding::UTF_8): string
{
    if ('' === $needle || null === search_ci($haystack, $needle, 0, $encoding)) {
        return $haystack;
    }

    return join(Regex\split($haystack, '{' . preg_quote($needle, '/') . '}iu'), $replacement);
}
```

```
1) Psl\Tests\Unit\Str\ReplaceCiTest::testReplaceCi with data set #2 ('مرحبا بكم', 'مرحبا سيف', 'سيف', 'بكم')
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'مرحبا بكم'
+'مرحبا '
```

As I'm debugging, I see that with the old `preg_split` returns the following array:

```php
array(2) {
  [0]=>
  string(11) "مرحبا "
  [1]=>
  string(0) ""
}
```

While calling `Psl\Regex` returns the following array, it looks like it is because of the `PREG_SPLIT_NO_EMPTY` flag:

```php
array(1) {
  [0]=>
  string(11) "مرحبا "
}
```

So in this case, if the dataset is correct, I will probably propose the following implementation:

```php
function replace_ci(string $haystack, string $needle, string $replacement, Encoding $encoding = Encoding::UTF_8): string
{
    if ('' === $needle || null === search_ci($haystack, $needle, 0, $encoding)) {
        return $haystack;
    }

    $pieces = Psl\Internal\suppress(
        static fn () => preg_split('{' . preg_quote($needle, '/') . '}iu', $haystack)
    );
   
    return join($pieces, $replacement);
}
```

WDYT @veewee  @azjezz?

Reply:
I don't know this language either. I'll keep it up to @azjezz

Reply:
sorry for the delayed response, the test is correct, and should pass.